0017-sampler-interface.md

```diff
@@ -0,0 +1,282 @@
+A bare bones sketch is below.
+Methods like `__or__` or static constructors from other formats are in-scope of the vision, but excluded for brevity.
+
+```python
+class BitArray(ShapedMixin):
+    """Stores bit-array outcomes."""
+
+    def __init__(self, array: NDArray, num_bits: int):
+        """
+        Args:
+            array: The data, where the last axis is over packed bits whose byte order is 
+                big-endian, the second last axis is over shots, and the preceding axes 
+                correspond to the shape of the experiment.
+            num_bits: How many bit are in each outcome; the size of the bit-array register.
+        """
+        self._array = np.array(array, copy=False, dtype=np.uint8)
+        if self._array.ndim < 2:
+            raise ValueError("The data arary needs to be at least 2-D")
+        self._num_bits = num_bits
+        if self._array.shape[-1] * 8 < self._num_Bits:
+            raise ValueError("The last axis of the data array is not big enough.")
+        # second last dimension is shots, last dimension is packed bits
+        self._shape = self._array.shape[:-2]
+
+    def __repr__(self):
+        desc = f"<num_samples={self.num_shots}, num_bits={self.num_bits}, shape={self.shape}>"
+        return f"BitArray({desc})"
+
+    @property
+    def array(self) -> NDArray[np.uint8]:
+        """The raw array of data."""
+        return self._array
+
+    @property
+    def num_bits(self) -> int:
+        """The number bits in each outcome."""
+        return self._num_bits
+
+    @property
+    def num_samples(self) -> int:
+        """The number of samples."""
+        return self._array.shape[-2]
+
+    def get_counts(self, loc: Tuple[int,...] | None = None, marginal_idxs: Tuple[int,...] | None = None) -> Counts:
+            def get_counts(self, idx: Tuple[int, ...] | None = None) -> Counts:
+        """Join all bit-array output registers together into a new counts structure.
+        
+        Args:
+            idx: Which task coordinate to fetch.
+
+        Raises:
+            ValueError: If no ``idx`` is given and this task result is more than 0-D.
+        """
+```
+
+### Measurement Levels
+
+`backend.run` has the notion of measurement levels which specify whether measurement instructions cause traces (not generallly implemented), complex phases, or discriminated bits to be returned.
+Eventually, measurement levels should be specified inside of a circuit itself by measuring into complex valued registers, as described in the OpenQASM 3 specification.
+This would effectively enable a mix-and-match between different measurement levels, and also allow a circuit to determine its own output types.
+
+However, before this becomes available, in order to acheive feature parity with `backend.run`, we propose implementing this as a flag (similar to `backend.run`) whose value changes the type of all bit-array outputs from `BitArray`s to NumPy arrays with `dtype=np.complex128`.
```

Comment:
Originally the plan was to have a different primitive, called "Discriminator", which would return IQ values for measurements (and then enable different options for how we do kernels etc. as a form of error suppression at this level). 
I think it's fine to fold this into the Sampler, especially since it becomes compatible with QASM3 and allows mix-and-match of different measurements, but maybe this should be called out that building a Discriminator is not planned anymore.

Reply:
Sure, though we never published a Qiskit roadmap toward other primitives.